staging/src/k8s.io/legacy-cloud-providers/aws/aws.go

```diff
@@ -1425,10 +1425,17 @@ func (c *Cloud) NodeAddresses(ctx context.Context, name types.NodeName) ([]v1.No
 				continue
 			}
 			ipPath := path.Join("network/interfaces/macs/", macID, "local-ipv4s")
-			macIPs[num], err = c.metadata.GetMetadata(ipPath)
+			v4ips, err := c.metadata.GetMetadata(ipPath)
 			if err != nil {
 				return nil, fmt.Errorf("error querying AWS metadata for %q: %q", ipPath, err)
 			}
+			ipPath = path.Join("network/interfaces/macs/", macID, "ipv6s")
```

Comment:
It seems like the EC2 meta makes a distinction for public vs private address for IPv4 (local-ipv4s vs public-ipv4), but not for IPv6 meaning there could be public addresses returned here but we are still storing these as `InternalIPs`. Am I missing something? cc @nckturner @M00nF1sh

Reply:
AFAIK ipv6 addresses in AWS are always public, assigned to the VPC

Reply:
That's good to know! In that case we should add this as type ExternalIP?

Reply:
Even though the IPv6 addresses are always public, in this context, I think we want them as kubernetes 'Internal' addresses for the kubelet.

Reply:
That means the apiserver could use a public v6 address over a private v4 address when talking to kubelet, that doesn't seem right unless I'm missing something.

Reply:
I believe this is the only IPv6 support AWS supports.

Reply:
I am under the impression that the firewall ACLs can be used to manage IPv6 access.

Reply:
I see, but if they are only public addresses they can still be added as type `ExternalIP` right?

Reply:
Clouds do goofy things with IPv4 addresses because they have to. They *don't* have to do those things with IPv6 addresses. There is no `InternalIP` / `ExternalIP` distinction for AWS IPv6 addresses. (If the apiserver uses a "public" v6 address to talk to the kubelet, then the packets will travel exactly the same route as they would if it used a private v4 address.)

In non-cloud environments, the node's IPs are considered `InternalIP` regardless of whether they are globally-routable or not, so it seems consistent to call public IPv6 addresses `InternalIP` as well. (ie, "internal" = "the node accepts traffic to this IP directly", "external" = "someone else redirects traffic from this IP to the node")

Reply:
I think that it makes sense use Internal based on Dan and Ryan comments,  and the fact that is how AWS works with IPv6 based on their docs

> For customers requiring a private subnet on their IPv6-enabled VPCs, we are introducing a new resource within the VPC called the Egress-only Internet Gateway, which can be setup to allow one-way access to Internet resources. With the Egress-only Internet Gateway, outgoing traffic to the Internet will be allowed. However, incoming traffic initiated from the Internet will be blocked

Also, IPv6 "ideal world" suggest using always public addresses ... we know the reality is other :)

Reply:
> In non-cloud environments, the node's IPs are considered InternalIP regardless of whether they are globally-routable or not, so it seems consistent to call public IPv6 addresses InternalIP as well. (ie, "internal" = "the node accepts traffic to this IP directly", "external" = "someone else redirects traffic from this IP to the node")

But this has generally been okay because we _only_ set `InternalIP` for non-cloud environments, so there's no priority check when there are multiple addresses. On AWS where there are multiple addresses of different types, if we always set IPv6 addresses as `InternalIP`, wouldn't we run the risk of apiserver preferring a public route over a private one? Even though the addresses are always "public" it sounds like we should check if they route publicly or privately before assigning a internal or external.

Reply:
In the AWS case, they route privately/directly. There's no IPv6 equivalent of the public IPv4 IPs.

Reply:
That's good to know, thanks for clarifying!